ipa-core/src/protocol/ipa_prf/shuffle/malicious.rs

```diff
@@ -0,0 +1,319 @@
+/// and sends `y1`, `c` to `H1` and `y2` to `H2`.
+///
+/// ## Errors
+/// Propagates network errors.
+async fn h3_verify<C: Context, S: BooleanArray>(
+    ctx: C,
+    keys: &[StdArray<Gf32Bit, 1>],
+    share_c_and_a: &[AdditiveShare<S>],
+    y1: &[S],
+    y2: &[S],
+) -> Result<(), Error> {
+    // compute hashes
+    // compute hash for y1
+    let hash_y1 = compute_row_hash::<S, _, _>(keys, y1);
+    // compute hash for y2
+    let hash_y2 = compute_row_hash::<S, _, _>(keys, y2);
+    // compute hash for C
+    let hash_c = compute_row_hash::<S, _, _>(
+        keys,
+        share_c_and_a.iter().map(ReplicatedSecretSharing::left),
+    );
+
+    // setup channels
+    let h1_ctx = ctx
+        .narrow(&OPRFShuffleStep::HashesH3toH1)
+        .set_total_records(TotalRecords::specified(2)?);
+    let h2_ctx = ctx
+        .narrow(&OPRFShuffleStep::HashH3toH2)
+        .set_total_records(TotalRecords::specified(1)?);
+    let channel_h1 = &h1_ctx.send_channel::<Hash>(ctx.role().peer(Direction::Right));
+    let channel_h2 = &h2_ctx.send_channel::<Hash>(ctx.role().peer(Direction::Left));
+
+    // send and receive hash
+    let _ = try_join3(
+        channel_h1.send(RecordId::FIRST, hash_y1),
+        channel_h1.send(RecordId::from(1usize), hash_c),
+        channel_h2.send(RecordId::FIRST, hash_y2),
+    )
+    .await?;
+
+    Ok(())
+}
+
+/// This function computes for each item in the iterator the inner product with `keys`.
+/// It concatenates all inner products and hashes them.
+///
+/// ## Panics
+/// Panics when conversion from `BooleanArray` to `Vec<Gf32Bit` fails.
+fn compute_row_hash<S, B, I>(keys: &[StdArray<Gf32Bit, 1>], row_iterator: I) -> Hash
+where
+    S: BooleanArray,
+    B: Borrow<S>,
+    I: IntoIterator<Item = B>,
+{
+    let iterator = row_iterator
+        .into_iter()
+        .map(|s| (*(s.borrow())).try_into().unwrap());
+    compute_hash(iterator.map(|row| {
+        row.iter()
+            .zip(keys)
+            .fold(Gf32Bit::ZERO, |acc, (row_entry, key)| {
+                acc + *row_entry * *key.first()
+            })
+    }))
+}
+
+/// This function reveals the MAC keys,
+/// stores them in a vector
+/// and appends a `Gf32Bit::ONE`
+///
+/// It uses `parallel_join` and therefore vector elements are a `StdArray` of length `1`.
+///
+/// ## Errors
+/// Propagates errors from `parallel_join` and `malicious_reveal`.
+async fn reveal_keys<C: Context>(
+    ctx: &C,
+    key_shares: &[AdditiveShare<Gf32Bit>],
+) -> Result<Vec<StdArray<Gf32Bit, 1>>, Error> {
+    // reveal MAC keys
+    let mut keys = ctx
+        .parallel_join(key_shares.iter().enumerate().map(|(i, key)| async move {
+            malicious_reveal(ctx.clone(), RecordId::from(i), None, key).await
+        }))
+        .await?
+        .into_iter()
+        .flatten()
+        .collect::<Vec<_>>();
+    // add a one, since last row element is tag which is not multiplied with a key
+    keys.push(iter::once(Gf32Bit::ONE).collect());
```

Comment:
this may cause reallocation, it should be better to collect everything at once like this

```rust
    // reveal MAC keys
    let keys = ctx
        .parallel_join(key_shares.iter().enumerate().map(|(i, key)| async move {
            malicious_reveal(ctx.clone(), RecordId::from(i), None, key).await
        }))
        .await?
        .into_iter()
        .flatten()
        // add a one, since last row element is tag which is not multiplied with a key
        .chain(iter::once(StdArray::from_fn(|_| Gf32Bit::ONE)))
        .collect::<Vec<_>>();

```